### PR TITLE
`no-loop-func` rule allows functions at init part (fixes #2427)

### LINKED ISSUE #2427: no-loop-func warnings at ForInStatement.right and ForStatement.init.
I got `no-loop-func` warnings at this code:

```js
"use strict";
let xs = [1, 2, 3];
for (let x of xs.filter(x0 => x0 != null)) {
  //                    ^^ here!
  console.log(x);
}
for (let x in xs.filter(x0 => x0 != null)) {
  //                    ^^ here!
  console.log(x);
}
for (let i = 0, ys = xs.filter(x0 => x0 != null); i < ys.length; i++) {
  //                           ^^ here!
  console.log(ys[i]);
}
```

But I think that `ForInStatement.right` and `ForStatement.init` are out of loop, since those run once at first.

I'm happy to make PR if this is accepted :)

----

version: `v0.20.0`
.eslintrc:

```
{
  "env": {
    "node": true,
    "es6": true
  },
  "ecmaFeatures": {
    "arrowFunctions": true,
    "forOf": true
  }
}
```

### PR DESCRIPTION
**Pull request by mysticatea**
This PR makes `no-loop-func` rule allow functions at init parts.

### CODE REVIEW

File: lib/rules/no-loop-func.js [hunk 1]
```diff
@@ -36,7 +19,31 @@ module.exports = function(context) {
...
         var ancestors = context.getAncestors();
 
-        if (ancestors.some(isLoop)) {
+        /**
+         * Checks if the given node is a loop and current context is in the loop.
+         * @param {ASTNode} ancestor The AST node to check.
+         * @param {number} index The index of ancestor in ancestors.
+         * @returns {boolean} Whether or not the node is a loop and current context is in the loop.
+         */
+        function hasInLoop(ancestor, index) {
```
Comment (nzakas) [inline lib/rules/no-loop-func.js]: I think `isInLoop` is a more accurate name.

**Comment by nzakas** [comment]
This looks really good. Just one small change.

Also, can you update your commit summary to begin with "Fix:"?

**Comment by mysticatea** [comment]
Thanks for the fast response!

I'll fix those.

**Comment by michaelficarra** [comment]
Just check the body. Functions could be in test and update.

**Comment by mysticatea** [comment]
Thanks for review!

I added two tests for `ForStatement.test` and `ForStatement.update`.

**Comment by nzakas** [comment]
@michaelficarra functions in test and update would fall prey to the same problem as those in body, would they not?

**Comment by michaelficarra** [comment]
Yeah, that's right.

**Comment by nzakas** [comment]
Okay cool. This looks good. Thanks @mysticatea

**Pull request merged**